dj/construction/build.py

```diff
@@ -1,184 +1,310 @@
-    tables: Dict[NodeRevision, List[ast.Table]] = {}
+    tables: DefaultDict[NodeRevision, List[ast.Table]] = collections.defaultdict(list)
 
     for table in select.find_all(ast.Table):
         if node := table.dj_node:  # pragma: no cover
-            tables[node] = tables.get(node, [])
             tables[node].append(table)
     return tables
 
 
-def _get_dim_cols_from_select(
-    select: ast.Select,
-) -> Dict[NodeRevision, List[ast.Column]]:
+def _join_path(
+    dimension_node: NodeRevision,
+    initial_nodes: Set[NodeRevision],
+) -> Tuple[NodeRevision, Dict[Tuple[NodeRevision, NodeRevision], List[Column]]]:
     """
-    Extract all dimension nodes referenced as columns
+    For a dimension node, we want to find a possible join path between it
+    and any of the nodes that are directly referenced in the original query. If
+    no join path exists, returns an empty dict.
     """
-    dimension_columns: Dict[NodeRevision, List[ast.Column]] = {}
+    processed = set()
+
+    to_process: Deque[
+        Tuple[NodeRevision, Dict[Tuple[NodeRevision], List[Column]]]
+    ] = collections.deque([])
+    join_info: Dict[Tuple[NodeRevision], List[Column]] = {}
+    to_process.extend([(node, join_info.copy()) for node in initial_nodes])
+
+    while to_process:
+        current_node, path = to_process.popleft()
+        processed.add(current_node)
+        dimensions_to_columns = collections.defaultdict(list)
+
+        # From the columns on the current node, find the next layer of
+        # dimension nodes that can be joined in
+        for col in current_node.columns:
+            if col.dimension:
+                dimensions_to_columns[col.dimension.current].append(col)
+
+        # Go through all potential dimensions and their join columns
+        for joinable_dim, join_cols in dimensions_to_columns.items():
+            next_join_path = {**path, **{(current_node, joinable_dim): join_cols}}
+            full_join_path = (joinable_dim, next_join_path)
+            if joinable_dim == dimension_node:
+                for col in join_cols:
+                    if col.dimension_column is None and not any(
+                        dim_col.name == "id" for dim_col in dimension_node.columns
+                    ):
+                        raise DJException(
+                            f"Node {current_node.name} specifying dimension "
+                            f"{joinable_dim.name} on column {col.name} does not"
+                            f" specify a dimension column, but {dimension_node.name} "
+                            f"does not have the default key `id`.",
+                        )
+                return full_join_path  # type: ignore
+            if joinable_dim not in processed:  # pragma: no cover
+                to_process.append(full_join_path)
+                for parent in joinable_dim.parents:
+                    to_process.append((parent.current, next_join_path))
+    return None  # type: ignore  # pragma: no cover
```

Comment:
Good point! I changed the logic here somewhat to find the shortest possible join path rather than just returning any path, so now there's no `return None` :)